tools/libSwiftSyntaxParser/CMakeLists.txt

```
@@ -1,61 +1,112 @@
 swift_swap_compiler_if_needed("SwiftSyntaxParser")
 
-# Use an 'internal' name, this is primarily intended for SwiftSyntax to import.
-set(SYNTAX_PARSER_LIB_NAME "_InternalSwiftSyntaxParser")
+function(create_parser_library target_name lib_name type)
+  set(SYNTAX_PARSER_LIB_NAME "${lib_name}")
 
-set(LLVM_EXPORTED_SYMBOL_FILE
-    ${CMAKE_CURRENT_SOURCE_DIR}/libSwiftSyntaxParser.exports)
+  set(LLVM_EXPORTED_SYMBOL_FILE
+      ${CMAKE_CURRENT_SOURCE_DIR}/libSwiftSyntaxParser.exports)
 
-add_swift_host_library(libSwiftSyntaxParser SHARED
-  c-include-check.c
-  libSwiftSyntaxParser.cpp
-  LLVM_LINK_COMPONENTS support)
-if(NOT SWIFT_BUILT_STANDALONE AND NOT CMAKE_C_COMPILER_ID MATCHES Clang)
-  add_dependencies(libSwiftSyntaxParser clang)
-endif()
-target_link_libraries(libSwiftSyntaxParser PRIVATE
-  swiftParse)
-set_target_properties(libSwiftSyntaxParser
-    PROPERTIES
-    OUTPUT_NAME ${SYNTAX_PARSER_LIB_NAME})
+  if ("${type}" STREQUAL relocatable_object_file)
+    # The -r linker option to produce a relocatable object file is not compatible with dead-stripping, so disable it.
+    set(LLVM_NO_DEAD_STRIP TRUE)
+    set(SWIFT_DISABLE_DEAD_STRIPPING TRUE)
+  endif()
+  add_swift_host_library("${target_name}" SHARED
+    c-include-check.c
+    libSwiftSyntaxParser.cpp
+    LLVM_LINK_COMPONENTS support)
+  if(NOT SWIFT_BUILT_STANDALONE AND NOT CMAKE_C_COMPILER_ID MATCHES Clang)
+    add_dependencies("${target_name}" clang)
+  endif()
+  target_link_libraries("${target_name}" PRIVATE
+    swiftParse)
+  set_target_properties("${target_name}"
+      PROPERTIES
+      OUTPUT_NAME ${SYNTAX_PARSER_LIB_NAME})
 
-add_llvm_symbol_exports(libSwiftSyntaxParser ${LLVM_EXPORTED_SYMBOL_FILE})
+  add_llvm_symbol_exports("${target_name}" ${LLVM_EXPORTED_SYMBOL_FILE})
 
-# Adds -dead_strip option
-add_link_opts(libSwiftSyntaxParser)
+  # Adds -dead_strip option
+  add_link_opts("${target_name}")
 
-if (SWIFT_LIBPARSER_VER)
-  set(SWIFTSYNTAX_PARSER_VERSION_STRING "${SWIFT_LIBPARSER_VER}")
-else()
-  set(SWIFTSYNTAX_PARSER_VERSION_STRING "${SWIFT_COMPILER_VERSION}")
-endif()
+  if ("${type}" STREQUAL relocatable_object_file)
+    # CMake does not support building static libraries that contain transitively referenced symbols by default.
+    # Perform the following workaround:
+    #  * Request a build of a shared library (aka. dylib) from CMake. This makes sure that all transitive dependencies are included in the linker invocation.
+    #  * Pass -r to the linker to produce a relocatable object file instead of a dylib
+    #  * Run libtool (below) to generate a static library from the relocatable object file, which we install into the toolchain.
+    
+    #  Explicitly link against C(++) libraries to make sure they end up in the final relocatable object file
+    target_compile_options("${target_name}" PRIVATE
+      "SHELL:-Xclang --linker-option=-lc++"
+      "SHELL:-Xclang --linker-option=-lncurses"
+      "SHELL:-Xclang --linker-option=-lz"
```

```suggestion
      "SHELL:-Xclang --dependent-lib=c++"
      "SHELL:-Xclang --dependent-lib=ncurses"
      "SHELL:-Xclang --dependent-lib=z"
```

This is a bit more clear semantically IMO, but otherwise is the same